Docs: Configuration file how-to guide (#10301)
readthedocs/readthedocs.org: docs/user/guides/setup/configuration-file.rst

Here are some variations of the configuration file that can help you get started.
Pick the one that resembles your project the most,
copy its contents to ``.readthedocs.yaml`` and add the file to your Git repository.

.. tabs::

    .. tab:: Sphinx

        If your project uses Sphinx,
        we offer a special builder optimized for Sphinx projects.

        .. literalinclude:: /config-file/examples/sphinx/.readthedocs.yaml
           :language: yaml
           :linenos:

    .. tab:: MkDocs

        If your project uses MkDocs,
        we offer a special builder optimized for MkDocs projects.

        .. literalinclude:: /config-file/examples/mkdocs/.readthedocs.yaml
           :language: yaml
           :linenos:

    .. tab:: Other tools

        If you are using another documentation tool,
        you can configure the commands of your documentation tool in the following code.

        You will need to create the output directories and direct your documentation tool to write its outputs into those directories.

        .. literalinclude:: /config-file/examples/custom-commands-echo/.readthedocs.yaml
           :language: yaml
           :linenos:

Editing the template

> --------------------

Reviewer: We should decide how this section should work.

Two alternatives:

1. Provide a list of sections that are of special interest for the new user (how it is now)
2. Simply provide guidance about the fact that the file has to be edited and then reference _the reference_.

Reviewer: If we go with option 1), we should add some more sections. Currently, we are just explaining two parts :) We should definitely also explain the `build.os` section.

Reviewer 2: Well, what I think we need to make our reference more usable is some better content _summarizing_ the configuration file.

This page will serve as a good place for the user to jump off into deeper content, and so I might still lean towards being more thorough with the number of configuration options we mention here, but not going into as much detail as we have.

For example, we describe the python key and what it is, and the python.install.requirements key and what it is below. Instead, I think I'd find the most use out of this page if it summarized the Python install options, but pointed to the relevant config reference for all details.

So, instead of the level of detail we have now, something closer to:

```rst
There are several configuration file options for installing your Python dependencies:

- Use :ref:`python.install.package` to install a package using pip
- Use :ref:`python.install.requirements` to install packages from a requirements file
- Use :ref:`build.jobs` to install packages using Poetry or PDM
```

Reviewer: Yup, I totally agree with that.

At the point of reading that section, the user should be familiar with changing content in the config file, so it should just be a question of introducing some important reference items.

Reviewer: Thanks for the copy-paste friendly suggestion :heart:

Reviewer 2: Oops, well I was writing in pseudo documentation, so I don't know how well it will paste :upside_down_face:

Reviewer: Yes but it's really nice with a clear example, copy-pasting and then adapting the example will be easier :)

Reviewer: Added a structure now. I found it hard to simply condense all possible adjustments in a simple list and opted for headings and `seealso::` boxes.

How to add a configuration file
===============================

> As part of the initial setup for your Read the Docs site,

Reviewer 2: ```suggestion
As part of the initial set up for your Read the Docs site,
```

you need to create a **configuration file** called ``.readthedocs.yaml``.
The configuration file tells Read the Docs what specific settings to use for your project.

This how-to guide covers:

> #. Where to put your configuration file.
> #. Templates to setup your configuration file from scratch.
> #. The most basic sections in the configuration file that can help you get started.
> #. Next steps...

Reviewer 2: This summary could use some condensing/clarification, so that it's more user-centric and the user knows exactly what they are getting from this guide:

```suggestion
#. Where to put your configuration file.
#. What to put in the configuration file.
#. How to customize the configuration for your project.
```

There is nuance there from our side of course, but this is what I'd care about as a user reading this guide

Reviewer: Sure - the previous list was a reflection of the section structure. Not a big loss, though.

.. seealso::

   :doc:`/tutorial/index`.
     Following the steps in our tutorial will help you setup your first documentation project.

Where to put your configuration file
------------------------------------

The ``.readthedocs.yaml`` file should be placed in the top-most directory of your project's repository.
We will get to the contents of the file in the next steps.

When you have changed the configuration file,
you need to commit and push the changes to your Git repository.
Read the Docs will then automatically find and use the configuration to build your project.

> .. note::
> 
>     Why is it called .yaml? What is that?
>       The configuration file is a `YAML`_ file. YAML files are a "map": a collection of
>       key-value pairs that can be nested. This is not unlike a JSON file or ``dict``
>       object in Python. This page won't explain the structure of YAML files, but many resources exist
>       online.
> 
> 
> .. _YAML: https://en.wikipedia.org/wiki/YAML

Reviewer 2: This section felt too low level for our docs. I think it's good to call out the fact that this is a YAML file, but we shouldn't need to explain YAML as a concept. If the user isn't familiar, they likely aren't familiar with most of the concepts/terms in this block anyways -- map, key-value paris, JSON, dict, etc.

To retain describing the structure without getting too deep into technical jargon, I think something close to this is best:

```suggestion
.. note::

    The Read the Docs configuration file is a `YAML`_ file. You can read more on the structure of this file and all of our configuration options :doc:`in our configuration file reference <>`.
```

Reviewer: I felt like this was a strong point in @nikblanchet's original howto: Cluing in the user so they don't feel like the guide is over their head if they haven't seen YAML before. You are not expected to have previous experience with YAML in order to follow this guide and copy the examples.

That's why the headline was: "Why is it called .yaml? What is that?" - to meet the user at their level.

I'm not sure about the change here: Do you still want to keep the reference to the Wikipedia article? I added that _again_ because I think it's great to flag that: No, you do not need to have met the concept of YAML before reading this guide, but here is a Wikipedia article explaining it.

Reviewer: If the note is just gonna be a cross-reference to the configuration file reference, I think we should delete it completely, already plenty of those :)

Reviewer 2: I do agree with the point about trying to be inclusive with this page, but I also feel familiarity with YAML is part of our base user persona. So, the copy/voice felt like it wasn't talking to a true user, to talking past our normal user -- it would feel the same as having a section explaining what Linux or what a terminal is to me.

I do agree on the cross reference, it isn't horribly important. I added it there to make the note more meaningful and pad out the content. I just updated my suggestion to remove this and link out to better resources. I meant to find a better link than Wikipedia, but agree that linking out is a good resource for this block.

For some comparison, here is the tutorial from CircleCI:
https://circleci.com/docs/config-intro/

It doesn't mention YAML in any detail, but does link out to a separate doc just for explaining YAML in more detail:
https://circleci.com/docs/introduction-to-yaml-configurations/

This is maybe a good addition for later for our docs, though I might argue we'd have trouble maintaining this if we're _just now_ adding a tutorial on our own configuration file :smile: 

From this page, and also from the YAML spec website, YAML is described as:

> YAML is a human-friendly data serialization language for all programming languages.

That's a good sweet spot for our description :dart:

I've updated the suggestion with this and linked to the YAML website. There is a link to the YAML spec, _but_ the overview and introduction are approachable and not frighteningly technical like specs normally are.

.. _howto_templates:

Getting started with a template
-------------------------------

> Here are some variations of the configuration file that can help you get started.

Reviewer 2: For some more definitive language and clarity:

```suggestion
Here are some configuration file examples to help you get started.
```

> Pick the one that resembles your project the most,

Reviewer 2: Some more exact language:

```suggestion
Pick an example based on the tool your project is using,
```

Reviewer: I added: Pick an example based on the tool **that** your project is using

copy its contents to ``.readthedocs.yaml`` and add the file to your Git repository.

.. tabs::

    .. tab:: Sphinx

        If your project uses Sphinx,
        we offer a special builder optimized for Sphinx projects.

        .. literalinclude:: /config-file/examples/sphinx/.readthedocs.yaml
           :language: yaml

>            :linenos:

Reviewer 2: ```suggestion
           :linenos:
           :caption: .readthedocs.yaml
```

    .. tab:: MkDocs

        If your project uses MkDocs,
        we offer a special builder optimized for MkDocs projects.

        .. literalinclude:: /config-file/examples/mkdocs/.readthedocs.yaml
           :language: yaml

>            :linenos:

Reviewer 2: ```suggestion
           :linenos:
           :caption: .readthedocs.yaml
```

    .. tab:: Other tools

        If you are using another documentation tool,
        you can configure the commands of your documentation tool in the following code.

        You will need to create the output directories and direct your documentation tool to write its outputs into those directories.

        .. literalinclude:: /config-file/examples/custom-commands-echo/.readthedocs.yaml
           :language: yaml
           :linenos:

Editing the template
--------------------

Now that you have a ``.readthedocs.yaml`` file added to your Git repository,

> you should see Read the Docs trying to build your project with the configuration file.

Reviewer 2: We actually don't make it obvious to users that the build is using a configuration file. We do surface this in the beta dashboard via the debug command output, but that's also the effective configuration file, not the actual file. It might not be clear to the user that the build is doing anything different.

We've talked a number of times of adding a debug flag to commands so we could output this level of information to users without it being noise, but that is still a ways out and probably only a beta dashboard thing still.

I'm not sure exactly what to say here though, other than "you should see the build doing something different" :thinking:

Reviewer: OH! Great point!! This is actually _a problem_. I just looked at a build, and no not even the raw log has this info. Can't we fix that in a separate issue very quickly? I'm not sure what's the best approach, but I imagine that this shouldn't have astronomic complexity? :)

Reviewer: Oh haha well, interesting timing you had for that input: https://github.com/readthedocs/readthedocs.org/pull/10413

Reviewer: Since this PR is already open, I think we can resolve this part of the how-to and try to align these two efforts.

Reviewer 2: Yeah, that makes sense, this is indeed a problem we should resolve. I think it's fine to mention this in our docs even if our UI doesn't exactly match.

The configuration file probably needs some adjustments to accommodate exactly your project setup.

.. note::

   If you added the configuration file in a separate branch,
   you may have to activate a :doc:`version </versions>` for that branch.

   If you have added the file in a pull request,
   you should enable :doc:`pull request builds </guides/pull-requests>`.

Skip: file header and comments
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

There are some parts of the templates that you can leave in place:

Comments

>   We added a lot of comments that explain stuff. They begin with a ``#``.

Reviewer 2: To reduce and use more exact language:

```suggestion
  We added comments that explain the configuration options and optional features.
  These lines begin with a ``#``.
```

Commented out features
  Similar to comments, we also use the ``#`` in front of some configuration options that may or may not be relevant.
  To enable something that's commented out,

>   simply delete the ``#`` and a blank space in front of the YAML code that you want to enable.

Reviewer 2: This feels similar to the YAML content above, where it feels like too much detail for our user personas. We're expecting our users to know Git and command line tooling, so I feel it's okay to assume they understand commenting.

It's good to explain the _why_ behind what we commented this out though. That does help the user I feel.

Reviewer: ```suggestion
Commented out features
  We use the ``#`` in front of some popular configuration options.
  They are there as examples,
  which you can choose to enable, delete or save for later.
```

Reviewer 2: That sounds great :+1:

``version`` key
  The version key tells the system how to read the rest of the configuration file.
  The current and only supported version is **version 2**.

Adjust: ``build.os``
~~~~~~~~~~~~~~~~~~~~

In our examples,
we are using Read the Docs' custom image based on the latest Ubuntu LTS release.
LTS means long-term support,

> meaning that your builds should not break within next many years.

Reviewer 2: Builds not breaking might be a lofty statement :smile: 

LTS might be the wrong focus here, I'm not sure. We're trying to tell the user that they won't need to change anything for several years by selecting an operating system version. We could skip on describing LTS here entirely, as this is more a security concept perhaps, and Ubuntu package freezing is the main benefit.

However,

> you should pay attention to this field if your current project needs to build on an older version of Ubuntu,

Reviewer 2: ```suggestion
you should pay attention to this field if your project needs to build on an older version of Ubuntu,
```

.. seealso::

   :ref:`config-file/v2:build.tools.python`
     Configuration file reference with all Python versions available for ``build.tools.python``.

   :ref:`config-file/v2:python`
     Configuration file reference for configuring the Python environment activated by ``build.tools.python``.

Adjust: Sphinx and MkDocs version
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

If you are using either the ``sphinx`` or ``mkdocs`` builder,
then Sphinx or MkDocs will be installed automatically in its latest version.

But we recommend that you specify the version that your documentation project uses.
The ``requirements`` key is a file path that points to a text (``.txt``) file
that lists the Python packages you want Read the Docs to install.

.. seealso::

   :ref:`guides/reproducible-builds:Use a requirements file for Python dependencies`
      This guide explains how to specify Python requirements,
      such as the version of Sphinx or MkDocs.

   :ref:`config-file/v2:sphinx`
     Configuration file reference for configuring the Sphinx builder.

   :ref:`config-file/v2:mkdocs`
     Configuration file reference for configuring the MkDocs builder.

.. _config_howto_build.tools:

> Adjust: Additional build environments?
> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer 2: To keep nomenclature consistent, probably "build tools" here, even though that is a term specific to RTD

```suggestion
Adjust: Additional build tools
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

> If you need software from additional platforms (nodejs, rust, golang),
> you can add and enable such an environment so you can install the packages you need.

Reviewer 2: More nomenclature consolidation:

```suggestion
If you need additional build tools (Node.js, Rust, Go),
you can define the versions of these tools to install as well, using the ``build.tools`` key.
```

.. seealso::

   :ref:`config-file/v2:build.tools`
     List of all tools that are possible to enable.

Adjust: Custom tools
~~~~~~~~~~~~~~~~~~~~

It's possible to build documentation using almost any documentation tool,
as long as an environment is available (see :ref:`the previous step <config_howto_build.tools>`).
Adding your own build commands is done by listing them in ``build.commands``.

The :ref:`Other tools <howto_templates>` template shows an example of how to do that.

.. seealso::

   :doc:`/build-customization`
     Jump to a full guide explaining how to setup (almost) any documentation tool.

   :ref:`config-file/v2:build.commands`
     Configuration file reference for custom build commands.

> 

Reviewer 2: I probably have the same feeling here, `build.commands` is not a good next step for novice users. The progression should try to balance giving the user control and us retaining some control, so: supported tool (sphinx/mkdocs) -> build jobs customization (intermediate, configuration file still works) -> build commands (expert, you're on your own).

Perhaps instead of jumping right to "custom tools" we should be pointing towards "custom jobs" and customizing the build process with new jobs. This at least serves us to promote advanced usage and make the user aware of things without through them into the gauntlet of build.commands.

Reviewer: I killed this section now!

I think the method that works is to have it all in a tab "Other tools" further up. So all this information should only be seen by users that aren't using Sphinx, MkDocs nor Docusaurus. It might be more _advanced_, but I think that it's better than leading them to a dead-end with no information available.

Reviewer 2: That's fair for now too. We do link out to the build customization page already, so perhaps that is what I am describing as well.

Next steps
----------

The configuration options that we mentioned in this guide aren't all.
There are more.

> Once you have your configuration file added to your Git repository,

Reviewer 2: To fix verb agreement with the rest of the sentence, and simplify a bit more:

```suggestion
After you add a configuration file your Git repository,
```
